### ISSUE TITLE
GLXBadDrawable Error when creating game window if use Nvidia prime offloading on linux

### ISSUE THREAD

**Issue by Alkaid-Benetnash** [Ryujinx/Ryujinx#2112]
### What's the issue you encountered?

When running Ryujinx with [nvidia prime offloading](https://download.nvidia.com/XFree86/Linux-x86_64/460.56/README/primerenderoffload.html) on linux, i.e. `prime-run Ryujinx`, the game table window works fine. But Gdk will report a "GLXBadDrawable" X Window System error then abort whenever I launch a game:
```
(Ryujinx:117270): Gdk-ERROR **: 15:32:55.010: The program 'Ryujinx' received an X Window System error.
This probably reflects a bug in the program.
The error was 'GLXBadDrawable'.
  (Details: serial 5118 error_code 160 request_code 152 (GLX) minor_code 5)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
/usr/bin/prime-run: line 2: 117270 Aborted                 (core dumped) __NV_PRIME_RENDER_OFFLOAD=1 __VK_LAYER_NV_optimus=NVIDIA_only __GLX_VENDOR_LIBRARY_NAME=nvidia "$@"
```
Without prime offloading and run X with nvidia gpu, then Ryujinx works fine.
The same issue has been reported on [reddit](https://www.reddit.com/r/Ryujinx/comments/i4x9ci/getting_an_error_with_nvidia_primerun_on_arch/) 7month ago

I did not make any changes to Ryujinx. The same problem happens with both Debug and Release build.
The version I tested is 88d0708.

I have pinpointed the error happens after executing the [`GLRenderer:OnConfigureEvent` call back](https://github.com/Ryujinx/Ryujinx/blob/88d07080612ad3dd12d99ddbdfee33406ea7c8be/Ryujinx/Ui/GLRenderer.cs#L242) and before [GLRenderWidget.Start()](https://github.com/Ryujinx/Ryujinx/blob/88d07080612ad3dd12d99ddbdfee33406ea7c8be/Ryujinx/Ui/MainWindow.cs#L643).

### How can the issue be reproduced?

1. `prime-run Ryujinx`
2. Double-click a game to launch it
3. The game window will not show up and Ryujinx will abort (in an unmanaged nvidia GLX library)

### Log file
The error happens in the unmanaged part, Ryujinx's log is not very helpful.
[Ryujinx_1.0.0-dirty_2021-03-15_15-32-44.log](https://github.com/Ryujinx/Ryujinx/files/6144200/Ryujinx_1.0.0-dirty_2021-03-15_15-32-44.log)
 So I also attach an lldb backtrace. I have debug symbols of libglib, libgtk. There is also a `clrstack` dotnet managed backtrace (from the dotnet-sos plugin) for your reference.
[lldb.log](https://github.com/Ryujinx/Ryujinx/files/6144209/lldb.log)

### Environment?

- Ryujinx version: 1.0.6785 or this commit 88d0708</br>
- Game version: Monster Hunter Rise Demo 1.0.2</br>
- System Specs:
  - OS: *Archlinux*
  - CPU: *i7-10750H*
  - GPU: *NVIDIA RTX 2070*
  - RAM: *24GB*
- Applied Mods : No

### Additional context?

Additional info about your environment:</br>
The nvidia prime is an official way to do gpu offloading so should be popular among laptops. It works well with other graphic emulators (e.g. rpcs3).
So I really appreciate someone can (help me) fix this.
I am happy to try things out or provide more logs. I do not know how to debug the GTK or GLX stack from here.

Thanks!

**Comment by edisionnano** [comment]
This is known and I think there already is another issue open.
Ryujinx doesn't work with prime (I haven't looked into why yet).
I suggest that you switch to the nvidia card entirely using optimus manager or similar software

**Comment by Alkaid-Benetnash** [comment]
FYI, it seems like the combination of ryujinx GLWidget + OpenTK3 is the problem.
OpenTK4 examples work with nvidia prime. So this problem is likely to be resolved once the project is updated to OpenTK4 #1561

**Comment by Stuntcokk87** [comment]
OpenTK4

On ***@***.*** Mar 27, 2021, 12:29 PM Alkaid
<notificGLWidget

**Comment by AcK77** [comment]
Could you test the build in https://github.com/Ryujinx/Ryujinx/pull/2194 and let us know if the issue is fixed with it?
Thanks!

**Comment by Alkaid-Benetnash** [comment]
Using the github bot artifact https://nightly.link/Ryujinx/Ryujinx/actions/artifacts/53007032.zip

There are two kinds of error messages:
1. Unhandled exception, [Ryujinx_1.0.0+a67d88b_2021-04-11_17-18-48.log](https://github.com/Ryujinx/Ryujinx/files/6292792/Ryujinx_1.0.0%2Ba67d88b_2021-04-11_17-18-48.log). Messages on terminal:
```
00:00:01.647 |I| Loader LoadExeFs: Loading subsdk0...
00:00:01.685 |I| Loader PrintRoSectionInfo: subsdk0:
    Module: multimedia
    SDK Libraries: SDK MW+Nintendo+NintendoSDK_movie-11_4_3-Release
00:00:01.685 |I| Loader LoadExeFs: Loading sdk...
00:00:01.767 |I| Loader PrintRoSectionInfo: sdk:
    Module: nnSdk
    FS SDK Version: 11.4.3
    SDK Libraries: SDK MW+Nintendo+NintendoSDK_libz-11_4_3-Release
                   SDK MW+Nintendo+NintendoSdk_nnSdk-11_4_3-Release
                   SDK MW+Nintendo+NintendoSDK_NVN-11_4_3-Release
00:00:01.805 |I| Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: True).
00:00:01.825 |I| Ptc Load: Loaded Profiling Info (size: 370144 bytes, profiled functions: 111431).
00:00:01.938 |I| Loader LoadNsos: Loading image 0 at 0x0000000008000000...
00:00:01.941 |I| Loader LoadNsos: Loading image 1 at 0x0000000008004000...
00:00:02.015 |I| Loader LoadNsos: Loading image 2 at 0x00000000157db000...
00:00:02.018 |I| Loader LoadNsos: Loading image 3 at 0x0000000015e81000...
00:00:02.027 |I| ModLoader LoadCheats: Build ids found for title 0100B04011742000:
    6F8B1226CC8CD958021D6BFD226EBF23F892AB9B000000000000000000000000
    6CDEC176EA1D12C5C4F85BF17213377C22D1EEEF000000000000000000000000
    507F19D1240D09D00992248E8A0FD77EBD784EB4000000000000000000000000
    2D369D9999E3E10E665106F840FFFF9BDAA0BB8E000000000000000000000000
00:00:02.027 |I| Loader LoadNca: Application Loaded: MONSTER HUNTER RISE v1.1.1 [0100b04011742000] [64-bit]
00:00:02.069 |I| GUI.WindowThread Hid Configure: Configured Controller ProController to Player1
XError: 0 result is 160
00:00:02.175 |E| Application : Unhandled exception caught: SPB.Graphics.Exceptions.ContextException: MakeCurrent() failed.
   at SPB.Platform.GLX.GLXOpenGLContext.MakeCurrent(NativeWindowBase window)
   at Ryujinx.Ui.GLWidget.Intialize() in /home/runner/work/Ryujinx/Ryujinx/Ryujinx/Ui/GLWidget.cs:line 83
   at Ryujinx.Ui.GLWidget.OnDrawn(Context cr) in /home/runner/work/Ryujinx/Ryujinx/Ryujinx/Ui/GLWidget.cs:line 47
   at Gtk.Widget.Drawn_cb(IntPtr inst, IntPtr cr)
```

2. abort + coredump, [Ryujinx_1.0.0+a67d88b_2021-04-11_17-19-07.log](https://github.com/Ryujinx/Ryujinx/files/6292795/Ryujinx_1.0.0%2Ba67d88b_2021-04-11_17-19-07.log). Messages on terminal:

```
00:00:01.836 |I| Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: True).
00:00:01.856 |I| Ptc Load: Loaded Profiling Info (size: 370144 bytes, profiled functions: 111431).
00:00:01.962 |I| Loader LoadNsos: Loading image 0 at 0x0000000008000000...
00:00:01.965 |I| Loader LoadNsos: Loading image 1 at 0x0000000008004000...
00:00:02.041 |I| Loader LoadNsos: Loading image 2 at 0x00000000157db000...
00:00:02.043 |I| Loader LoadNsos: Loading image 3 at 0x0000000015e81000...
00:00:02.053 |I| ModLoader LoadCheats: Build ids found for title 0100B04011742000:
    6F8B1226CC8CD958021D6BFD226EBF23F892AB9B000000000000000000000000
    6CDEC176EA1D12C5C4F85BF17213377C22D1EEEF000000000000000000000000
    507F19D1240D09D00992248E8A0FD77EBD784EB4000000000000000000000000
    2D369D9999E3E10E665106F840FFFF9BDAA0BB8E000000000000000000000000
00:00:02.053 |I| Loader LoadNca: Application Loaded: MONSTER HUNTER RISE v1.1.1 [0100b04011742000] [64-bit]
00:00:02.095 |I| GUI.WindowThread Hid Configure: Configured Controller ProController to Player1
Process terminated. A callback was made on a garbage collected delegate of type 'SPB!SPB.Platform.X11.X11+XErrorHandler::Invoke'.
Repeat 2 times:
--------------------------------
   at SPB.Platform.GLX.GLX.MakeCurrent(IntPtr, IntPtr, IntPtr)
--------------------------------
   at SPB.Platform.GLX.GLXOpenGLContext.MakeCurrent(SPB.Windowing.NativeWindowBase)
   at Ryujinx.Ui.GLWidget.Intialize()
   at Ryujinx.Ui.GLWidget.OnDrawn(Cairo.Context)
   at Gtk.Widget.Drawn_cb(IntPtr, IntPtr)
   at Ryujinx.Program.Main(System.String[])
./run_test.sh: line 3:  3576 Aborted                 (core dumped) ${RELEASE}/Ryujinx -r `dirname ${0}`/data/test.data
```

**Comment by Thog** [comment]
> Using the github bot artifact [ryujinx-Release-1.0.0+a67d88b-linux_x64](https://nightly.link/Ryujinx/Ryujinx/actions/artifacts/53007032.zip)
> 
> There are two kinds of error messages:
> 
>     1. Unhandled exception, [Ryujinx_1.0.0+a67d88b_2021-04-11_17-18-48.log](https://github.com/Ryujinx/Ryujinx/files/6292792/Ryujinx_1.0.0%2Ba67d88b_2021-04-11_17-18-48.log). Messages on terminal:
> 
> 
> ```
> 00:00:01.647 |I| Loader LoadExeFs: Loading subsdk0...
> 00:00:01.685 |I| Loader PrintRoSectionInfo: subsdk0:
>     Module: multimedia
>     SDK Libraries: SDK MW+Nintendo+NintendoSDK_movie-11_4_3-Release
> 00:00:01.685 |I| Loader LoadExeFs: Loading sdk...
> 00:00:01.767 |I| Loader PrintRoSectionInfo: sdk:
>     Module: nnSdk
>     FS SDK Version: 11.4.3
>     SDK Libraries: SDK MW+Nintendo+NintendoSDK_libz-11_4_3-Release
>                    SDK MW+Nintendo+NintendoSdk_nnSdk-11_4_3-Release
>                    SDK MW+Nintendo+NintendoSDK_NVN-11_4_3-Release
> 00:00:01.805 |I| Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: True).
> 00:00:01.825 |I| Ptc Load: Loaded Profiling Info (size: 370144 bytes, profiled functions: 111431).
> 00:00:01.938 |I| Loader LoadNsos: Loading image 0 at 0x0000000008000000...
> 00:00:01.941 |I| Loader LoadNsos: Loading image 1 at 0x0000000008004000...
> 00:00:02.015 |I| Loader LoadNsos: Loading image 2 at 0x00000000157db000...
> 00:00:02.018 |I| Loader LoadNsos: Loading image 3 at 0x0000000015e81000...
> 00:00:02.027 |I| ModLoader LoadCheats: Build ids found for title 0100B04011742000:
>     6F8B1226CC8CD958021D6BFD226EBF23F892AB9B000000000000000000000000
>     6CDEC176EA1D12C5C4F85BF17213377C22D1EEEF000000000000000000000000
>     507F19D1240D09D00992248E8A0FD77EBD784EB4000000000000000000000000
>     2D369D9999E3E10E665106F840FFFF9BDAA0BB8E000000000000000000000000
> 00:00:02.027 |I| Loader LoadNca: Application Loaded: MONSTER HUNTER RISE v1.1.1 [0100b04011742000] [64-bit]
> 00:00:02.069 |I| GUI.WindowThread Hid Configure: Configured Controller ProController to Player1
> XError: 0 result is 160
> 00:00:02.175 |E| Application : Unhandled exception caught: SPB.Graphics.Exceptions.ContextException: MakeCurrent() failed.
>    at SPB.Platform.GLX.GLXOpenGLContext.MakeCurrent(NativeWindowBase window)
>    at Ryujinx.Ui.GLWidget.Intialize() in /home/runner/work/Ryujinx/Ryujinx/Ryujinx/Ui/GLWidget.cs:line 83
>    at Ryujinx.Ui.GLWidget.OnDrawn(Context cr) in /home/runner/work/Ryujinx/Ryujinx/Ryujinx/Ui/GLWidget.cs:line 47
>    at Gtk.Widget.Drawn_cb(IntPtr inst, IntPtr cr)
> ```
> 
>     1. abort + coredump, [Ryujinx_1.0.0+a67d88b_2021-04-11_17-19-07.log](https://github.com/Ryujinx/Ryujinx/files/6292795/Ryujinx_1.0.0%2Ba67d88b_2021-04-11_17-19-07.log). Messages on terminal:
> 
> 
> ```
> 00:00:01.836 |I| Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: True).
> 00:00:01.856 |I| Ptc Load: Loaded Profiling Info (size: 370144 bytes, profiled functions: 111431).
> 00:00:01.962 |I| Loader LoadNsos: Loading image 0 at 0x0000000008000000...
> 00:00:01.965 |I| Loader LoadNsos: Loading image 1 at 0x0000000008004000...
> 00:00:02.041 |I| Loader LoadNsos: Loading image 2 at 0x00000000157db000...
> 00:00:02.043 |I| Loader LoadNsos: Loading image 3 at 0x0000000015e81000...
> 00:00:02.053 |I| ModLoader LoadCheats: Build ids found for title 0100B04011742000:
>     6F8B1226CC8CD958021D6BFD226EBF23F892AB9B000000000000000000000000
>     6CDEC176EA1D12C5C4F85BF17213377C22D1EEEF000000000000000000000000
>     507F19D1240D09D00992248E8A0FD77EBD784EB4000000000000000000000000
>     2D369D9999E3E10E665106F840FFFF9BDAA0BB8E000000000000000000000000
> 00:00:02.053 |I| Loader LoadNca: Application Loaded: MONSTER HUNTER RISE v1.1.1 [0100b04011742000] [64-bit]
> 00:00:02.095 |I| GUI.WindowThread Hid Configure: Configured Controller ProController to Player1
> Process terminated. A callback was made on a garbage collected delegate of type 'SPB!SPB.Platform.X11.X11+XErrorHandler::Invoke'.
> Repeat 2 times:
> --------------------------------
>    at SPB.Platform.GLX.GLX.MakeCurrent(IntPtr, IntPtr, IntPtr)
> --------------------------------
>    at SPB.Platform.GLX.GLXOpenGLContext.MakeCurrent(SPB.Windowing.NativeWindowBase)
>    at Ryujinx.Ui.GLWidget.Intialize()
>    at Ryujinx.Ui.GLWidget.OnDrawn(Cairo.Context)
>    at Gtk.Widget.Drawn_cb(IntPtr, IntPtr)
>    at Ryujinx.Program.Main(System.String[])
> ./run_test.sh: line 3:  3576 Aborted                 (core dumped) ${RELEASE}/Ryujinx -r `dirname ${0}`/data/test.data
> ```
> 
> These errors happened after clicking the game. The main UI window (game list + menu) works fine. Such behavior is the same as before.

I tested explicitly with prime-run and without on Linux when working on it and didn't got anything like that (CPU: i7-10750H, GPU: RTX 2070 on Arch)

Tho for the second error, that's my bad on this one (the handler of X errors being freed)

**Comment by Alkaid-Benetnash** [comment]
@Thog I have basically the same spec. Arch, prime-run, Intel 10gen i5, 2070.
These two errors seem to share the root cause die to similar call stack. I suppose the handler free you talked about leads to undefined behavior.

Should I wait for yourvfix of the X errors and try again?

**Comment by edisionnano** [comment]
I'm 99% sure the issue is fixed now. Someone please try. If you don't have prime-run just do `__GLX_VENDOR_LIBRARY_NAME=nvidia DRI_PRIME=1 ./Ryujinx`

**Comment by ctr305** [comment]
> I'm 99% sure the issue is fixed now. Someone please try. If you don't have prime-run just do `__GLX_VENDOR_LIBRARY_NAME=nvidia DRI_PRIME=1 ./Ryujinx`

Not quite, at least not form me.
Opening any game with prime-run or the command you provided results in the program crashing with the following console log:
```
00:00:00.002 |W| Application GetWindowScaleFactor: Couldn't determine monitor DPI: The type initializer for 'Gdip' threw an exception.
00:00:00.297 |I| Configuration LogValueChange: EnableFileLog set to: True
00:00:00.322 |I| Configuration LogValueChange: BackendThreading set to: Auto
00:00:00.323 |I| Configuration LogValueChange: ResScale set to: 1
00:00:00.325 |I| Configuration LogValueChange: ResScaleCustom set to: 1
00:00:00.325 |I| Configuration LogValueChange: MaxAnisotropy set to: -1
00:00:00.326 |I| Configuration LogValueChange: AspectRatio set to: Fixed16x9
00:00:00.331 |I| Configuration LogValueChange: EnableDockedMode set to: False
00:00:00.332 |I| Configuration LogValueChange: EnableVsync set to: True
00:00:00.332 |I| Configuration LogValueChange: EnableShaderCache set to: True
00:00:00.332 |I| Configuration LogValueChange: EnablePtc set to: True
00:00:00.333 |I| Configuration LogValueChange: EnableFsIntegrityChecks set to: True
00:00:00.333 |I| Configuration LogValueChange: FsGlobalAccessLogMode set to: 0
00:00:00.334 |I| Configuration LogValueChange: AudioBackend set to: SDL2
00:00:00.335 |I| Configuration LogValueChange: MemoryManagerMode set to: HostMappedUnsafe
00:00:00.335 |I| Configuration LogValueChange: ExpandRam set to: False
00:00:00.336 |I| Configuration LogValueChange: IgnoreMissingServices set to: False
00:00:00.342 |N| Application PrintSystemInfo: Ryujinx Version: 1.0.7101
00:00:00.349 |N| Application Print: Operating System: Linux 5.14.16-zen1-1-zen #1 ZEN SMP PREEMPT Tue, 02 Nov 2021 22:23:00 +0000 (X64)
00:00:00.349 |N| Application Print: CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz ; 8 logical
00:00:00.349 |N| Application Print: RAM: Total 15880 MB ; Available 9199 MB
00:00:00.353 |N| Application PrintSystemInfo: Logs Enabled: Info, Warning, Error, Guest, Stub
00:00:00.353 |N| Application PrintSystemInfo: Launch Mode: UserProfile
00:00:00.123 |I| Gpu : Backend Threading (Auto): True
00:00:00.569 |N| Application LoadApplication: Using Firmware Version: 13.0.0
00:00:00.569 |I| Application LoadApplication: Loading as NSP.
00:00:00.626 |I| ModLoader QueryContentsDir: Searching mods for Title 010093801237C000
00:00:00.637 |I| Application EnsureSaveData: Ensuring required savedata exists.
00:00:00.651 |I| Loader LoadExeFs: Loading rtld...
00:00:00.754 |I| Loader PrintRoSectionInfo: rtld:
    Module: nnrtld
00:00:00.754 |I| Loader LoadExeFs: Loading main...
00:00:00.977 |I| Loader PrintRoSectionInfo: main:
    Module: C:\perforcelocal\[ws00]\PRJ_04\GAME\Code\_output\NX64\apps\gameapp\Retail\cazadora.nss
    SDK Libraries: SDK MW+Nintendo+NintendoWare_Atk-11_4_3-Release
00:00:00.977 |I| Loader LoadExeFs: Loading subsdk0...
00:00:01.028 |I| Loader PrintRoSectionInfo: subsdk0:
    Module: multimedia
    SDK Libraries: SDK MW+Nintendo+NintendoSDK_movie-11_4_3-Release
00:00:01.028 |I| Loader LoadExeFs: Loading sdk...
00:00:01.129 |I| Loader PrintRoSectionInfo: sdk:
    Module: nnSdk
    FS SDK Version: 11.4.3
    SDK Libraries: SDK MW+Nintendo+NintendoSDK_libz-11_4_3-Release
                   SDK MW+Nintendo+NintendoSdk_nnSdk-11_4_3-Release
                   SDK MW+Nintendo+NintendoSDK_NVN-11_4_3-Release
00:00:01.189 |I| Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: True).
00:00:01.224 |I| Ptc Load: Loaded Profiling Info (size: 94293 bytes, profiled functions: 29864).
00:00:01.466 |I| Loader LoadNsos: Loading image 0 at 0x0000000008000000...
00:00:01.469 |I| Loader LoadNsos: Loading image 1 at 0x0000000008004000...
00:00:02.306 |I| Loader LoadNsos: Loading image 2 at 0x000000007a9c1000...
00:00:02.310 |I| Loader LoadNsos: Loading image 3 at 0x000000007b067000...
00:00:02.322 |I| ModLoader LoadCheats: Build ids found for title 010093801237C000:
    6F8B1226CC8CD958021D6BFD226EBF23F892AB9B000000000000000000000000
    49161D9CCBC15DF944D0B6278A3C446C006B0BE8000000000000000000000000
    507F19D1240D09D00992248E8A0FD77EBD784EB4000000000000000000000000
    2D369D9999E3E10E665106F840FFFF9BDAA0BB8E000000000000000000000000
00:00:02.322 |I| Loader LoadNca: Application Loaded: Metroid Dread v1.0.0 [010093801237c000] [64-bit]
00:00:02.331 |I| GUI.WindowThread Hid Configure: Configured Controller ProController to Player1
XError: 0 result is 2
XError: 0 result is 167
XError: 0 result is 167
00:00:02.416 |E| Application : Unhandled exception caught: SPB.Graphics.Exceptions.ContextException: CreateContext() failed.
   at SPB.Platform.GLX.GLXOpenGLContext.Initialize(NativeWindowBase window)
   at Ryujinx.Ui.GlRenderer.IntializeOpenGL()
   at Ryujinx.Ui.GlRenderer.OnDrawn(Context cr)
   at Gtk.Widget.Drawn_cb(IntPtr inst, IntPtr cr)
```

Tested with Metroid Dread on a i7-7700hq and a GTX 1050 Mobile (GP107M) running EndeavourOS 64 bit with kernel 5.14.16-zen1-1-zen and proprietary nvidia driver version 495.44

**Comment by CeruleanSky** [comment]
@ctr305 There are sdl2-headless and avalonia builds in the below comment, could try those

https://github.com/Ryujinx/Ryujinx/pull/2507#issuecomment-886876402

**Comment by ctr305** [comment]
@CeruleanSky 

Just got done testing them out on the same setup.

- ryujinx-headless-sdl2-Release-1.0.0+15beb39-linux_x64: Works perfectly with prime-run!
- ryujinx-ava-Release-1.0.0+15beb39-linux_x64: Same issue as normal build. Crashes upon load, log included with this comment.
[Ryujinx_1.0.0+15beb39_2021-11-09_13-29-25.log](https://github.com/Ryujinx/Ryujinx/files/7507468/Ryujinx_1.0.0%2B15beb39_2021-11-09_13-29-25.log)

- ryujinx-Release-1.0.0+15beb39-linux_x64: Game loads and audio can be heard playing, but window displays nothing but a frozen frame of the main UI. log included with this comment.
[Ryujinx_1.0.0+15beb39_2021-11-09_13-18-31.log](https://github.com/Ryujinx/Ryujinx/files/7507457/Ryujinx_1.0.0%2B15beb39_2021-11-09_13-18-31.log)

**Comment by edisionnano** [comment]
> > I'm 99% sure the issue is fixed now. Someone please try. If you don't have prime-run just do `__GLX_VENDOR_LIBRARY_NAME=nvidia DRI_PRIME=1 ./Ryujinx`
> 
> Not quite, at least not form me. Opening any game with prime-run or the command you provided results in the program crashing with the following console log:
> 
> ```
> 00:00:00.002 |W| Application GetWindowScaleFactor: Couldn't determine monitor DPI: The type initializer for 'Gdip' threw an exception.
> 00:00:00.297 |I| Configuration LogValueChange: EnableFileLog set to: True
> 00:00:00.322 |I| Configuration LogValueChange: BackendThreading set to: Auto
> 00:00:00.323 |I| Configuration LogValueChange: ResScale set to: 1
> 00:00:00.325 |I| Configuration LogValueChange: ResScaleCustom set to: 1
> 00:00:00.325 |I| Configuration LogValueChange: MaxAnisotropy set to: -1
> 00:00:00.326 |I| Configuration LogValueChange: AspectRatio set to: Fixed16x9
> 00:00:00.331 |I| Configuration LogValueChange: EnableDockedMode set to: False
> 00:00:00.332 |I| Configuration LogValueChange: EnableVsync set to: True
> 00:00:00.332 |I| Configuration LogValueChange: EnableShaderCache set to: True
> 00:00:00.332 |I| Configuration LogValueChange: EnablePtc set to: True
> 00:00:00.333 |I| Configuration LogValueChange: EnableFsIntegrityChecks set to: True
> 00:00:00.333 |I| Configuration LogValueChange: FsGlobalAccessLogMode set to: 0
> 00:00:00.334 |I| Configuration LogValueChange: AudioBackend set to: SDL2
> 00:00:00.335 |I| Configuration LogValueChange: MemoryManagerMode set to: HostMappedUnsafe
> 00:00:00.335 |I| Configuration LogValueChange: ExpandRam set to: False
> 00:00:00.336 |I| Configuration LogValueChange: IgnoreMissingServices set to: False
> 00:00:00.342 |N| Application PrintSystemInfo: Ryujinx Version: 1.0.7101
> 00:00:00.349 |N| Application Print: Operating System: Linux 5.14.16-zen1-1-zen #1 ZEN SMP PREEMPT Tue, 02 Nov 2021 22:23:00 +0000 (X64)
> 00:00:00.349 |N| Application Print: CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz ; 8 logical
> 00:00:00.349 |N| Application Print: RAM: Total 15880 MB ; Available 9199 MB
> 00:00:00.353 |N| Application PrintSystemInfo: Logs Enabled: Info, Warning, Error, Guest, Stub
> 00:00:00.353 |N| Application PrintSystemInfo: Launch Mode: UserProfile
> 00:00:00.123 |I| Gpu : Backend Threading (Auto): True
> 00:00:00.569 |N| Application LoadApplication: Using Firmware Version: 13.0.0
> 00:00:00.569 |I| Application LoadApplication: Loading as NSP.
> 00:00:00.626 |I| ModLoader QueryContentsDir: Searching mods for Title 010093801237C000
> 00:00:00.637 |I| Application EnsureSaveData: Ensuring required savedata exists.
> 00:00:00.651 |I| Loader LoadExeFs: Loading rtld...
> 00:00:00.754 |I| Loader PrintRoSectionInfo: rtld:
>     Module: nnrtld
> 00:00:00.754 |I| Loader LoadExeFs: Loading main...
> 00:00:00.977 |I| Loader PrintRoSectionInfo: main:
>     Module: C:\perforcelocal\[ws00]\PRJ_04\GAME\Code\_output\NX64\apps\gameapp\Retail\cazadora.nss
>     SDK Libraries: SDK MW+Nintendo+NintendoWare_Atk-11_4_3-Release
> 00:00:00.977 |I| Loader LoadExeFs: Loading subsdk0...
> 00:00:01.028 |I| Loader PrintRoSectionInfo: subsdk0:
>     Module: multimedia
>     SDK Libraries: SDK MW+Nintendo+NintendoSDK_movie-11_4_3-Release
> 00:00:01.028 |I| Loader LoadExeFs: Loading sdk...
> 00:00:01.129 |I| Loader PrintRoSectionInfo: sdk:
>     Module: nnSdk
>     FS SDK Version: 11.4.3
>     SDK Libraries: SDK MW+Nintendo+NintendoSDK_libz-11_4_3-Release
>                    SDK MW+Nintendo+NintendoSdk_nnSdk-11_4_3-Release
>                    SDK MW+Nintendo+NintendoSDK_NVN-11_4_3-Release
> 00:00:01.189 |I| Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: True).
> 00:00:01.224 |I| Ptc Load: Loaded Profiling Info (size: 94293 bytes, profiled functions: 29864).
> 00:00:01.466 |I| Loader LoadNsos: Loading image 0 at 0x0000000008000000...
> 00:00:01.469 |I| Loader LoadNsos: Loading image 1 at 0x0000000008004000...
> 00:00:02.306 |I| Loader LoadNsos: Loading image 2 at 0x000000007a9c1000...
> 00:00:02.310 |I| Loader LoadNsos: Loading image 3 at 0x000000007b067000...
> 00:00:02.322 |I| ModLoader LoadCheats: Build ids found for title 010093801237C000:
>     6F8B1226CC8CD958021D6BFD226EBF23F892AB9B000000000000000000000000
>     49161D9CCBC15DF944D0B6278A3C446C006B0BE8000000000000000000000000
>     507F19D1240D09D00992248E8A0FD77EBD784EB4000000000000000000000000
>     2D369D9999E3E10E665106F840FFFF9BDAA0BB8E000000000000000000000000
> 00:00:02.322 |I| Loader LoadNca: Application Loaded: Metroid Dread v1.0.0 [010093801237c000] [64-bit]
> 00:00:02.331 |I| GUI.WindowThread Hid Configure: Configured Controller ProController to Player1
> XError: 0 result is 2
> XError: 0 result is 167
> XError: 0 result is 167
> 00:00:02.416 |E| Application : Unhandled exception caught: SPB.Graphics.Exceptions.ContextException: CreateContext() failed.
>    at SPB.Platform.GLX.GLXOpenGLContext.Initialize(NativeWindowBase window)
>    at Ryujinx.Ui.GlRenderer.IntializeOpenGL()
>    at Ryujinx.Ui.GlRenderer.OnDrawn(Context cr)
>    at Gtk.Widget.Drawn_cb(IntPtr inst, IntPtr cr)
> ```
> 
> Tested with Metroid Dread on a i7-7700hq and a GTX 1050 Mobile (GP107M) running EndeavourOS 64 bit with kernel 5.14.16-zen1-1-zen and proprietary nvidia driver version 495.44 on Ryujinx version 1.0.7101-linux_x64

Does the window launch at all? Or does the window crash when entering the game?

**Comment by CeruleanSky** [comment]
@ctr305 

For the avalonia build your log says:
` ---> System.DllNotFoundException: Unable to load shared library 'libgdiplus' or one of its dependencies.`

Perhaps you can install libgdiplus with pacman or verify it is installed [Problem: GDIPlusInit](https://www.mono-project.com/docs/gui/problemgdiplusinit/)

**Comment by ctr305** [comment]
@CeruleanSky 

After installing the library, the game now launches and the audio can be heard, but it displays nothing but a black screen. Thank you for the observation
![Screenshot_20211109_154145](https://user-images.githubusercontent.com/84721138/141009665-db3ada4c-debe-43ad-8e53-296a908db1ba.png)
.

log: 
[Ryujinx_1.0.0+15beb39_2021-11-09_15-40-38.log](https://github.com/Ryujinx/Ryujinx/files/7508171/Ryujinx_1.0.0%2B15beb39_2021-11-09_15-40-38.log)

@edisionnano 

The emulator crashes immediately upon loading a game and then exits, but the main menu and the option menus work as expected.

**Comment by TanapTheTimid** [comment]
I'm facing the same issue on Arch linux nvidia driver version 470.94. The latest avalonia build works perfectly, but neither the official build nor the headless-sdl2 build works.

**Comment by edisionnano** [comment]
For anyone watching the issue, this issue will be resolved once EGL is merged. It will also give us a bit better performance. For anyone who wants to try here's a build with EGL merged.
https://cdn.discordapp.com/attachments/790357964749406249/934871942694047754/publish.zip
This build is built before the fix for the pokemon saves were merged so if you have pokemon saves don't use it as it will corrupt them.

**Comment by username227** [comment]
I am having the same issue here. Any timeframe regarding this EGL merging? Thanks so much for all that you guys do.

**Comment by EeveeEuphoria** [comment]
> @username227 if all else fails get the Avalonia build, it should not happen there.

I've compiled the Avalonia build myself ([using this branch](https://github.com/emmauss/Ryujinx/tree/avalonia), since that's the one that's linked), and I still get the same issue that you closed my issue as duplicate for.

> For anyone watching the issue, this issue will be resolved once EGL is merged. It will also give us a bit better performance. For anyone who wants to try here's a build with EGL merged. https://cdn.discordapp.com/attachments/790357964749406249/934871942694047754/publish.zip This build is built before the fix for the pokemon saves were merged so if you have pokemon saves don't use it as it will corrupt them.

However, using this build seems to work, for some reason? I'm not sure if I've gotten the wrong merge, but I got the implication that Avalonia has this merge.

**Comment by edisionnano** [comment]
> > @username227 if all else fails get the Avalonia build, it should not happen there.
> 
> I've compiled the Avalonia build myself ([using this branch](https://github.com/emmauss/Ryujinx/tree/avalonia), since that's the one that's linked), and I still get the same issue that you closed my issue as duplicate for.
> 
> > For anyone watching the issue, this issue will be resolved once EGL is merged. It will also give us a bit better performance. For anyone who wants to try here's a build with EGL merged. https://cdn.discordapp.com/attachments/790357964749406249/934871942694047754/publish.zip This build is built before the fix for the pokemon saves were merged so if you have pokemon saves don't use it as it will corrupt them.
> 
> However, using this build seems to work, for some reason? I'm not sure if I've gotten the wrong merge, but I got the implication that Avalonia has this merge.
The avalonia build should work fine, use the prebuilt builds the bot provided on the comments of the avalonia pr.

I don't know what you mean by merge but you shouldn't merge anything. Avalonia doesn't support EGL for OpenGL.

**Comment by EeveeEuphoria** [comment]
> I don't know what you mean by merge but you shouldn't merge anything. Avalonia doesn't support EGL for OpenGL.

By merge I was just referring to Avalonia. I wanted to compile it myself to see if it would work that way, but it didn't, so yeah. The build you provided works however.

**Comment by Mario-119** [comment]
Any updates on this? It seems that even disabling the iGPU still causes a crash to happen, on Mint 20.3 for reference.

**Comment by Leopard1907** [comment]
Hitting the same issue here with Arch Linux on Optimus/Prime laptop.

```
XError: 0 result is 160
00:00:02.407 |E| Application : Unhandled exception caught: SPB.Graphics.Exceptions.ContextException: MakeCurrent() failed.
   at SPB.Platform.GLX.GLXOpenGLContext.MakeCurrent(NativeWindowBase window)
   at Ryujinx.Ui.GlRenderer.IntializeOpenGL() in D:\a\Ryujinx\Ryujinx\Ryujinx\Ui\GLRenderer.cs:line 50
   at Ryujinx.Ui.GlRenderer.OnDrawn(Context cr) in D:\a\Ryujinx\Ryujinx\Ryujinx\Ui\GLRenderer.cs:line 36
   at Gtk.Widget.Drawn_cb(IntPtr inst, IntPtr cr)
   ```

**Comment by mdnrealdeal** [comment]
Hello, for those still having issues on PRIME laptops on Linux, there is a workaround that I have found which works for now (at least for me.) You can use the Vulkan backend PR (located at #2518) and it seems to work fine for me in PRIME offloading.

**Comment by emmauss** [comment]
> Hello, for those still having issues on PRIME laptops on Linux, there is a workaround that I have found which works for now (at least for me.) You can use the Vulkan backend PR (located at #2518) and it seems to work fine for me in PRIME offloading. This is a workaround for now, if you can't make the Avalonia branch work with your system. Hope this helps all that reads this!

The reason vulkan works as a workaround is because vulkan doesn't use glx, and everything needed to draw is configured by the user in the swap chain. Avalonia build works because before the window is created, a fb config that's compatible with your discrete gpu is requested and that is used to initialize the window. Gtk doesn't request one that's compatible our context and the gpu, so it crashes when trying to make current.

**Comment by marcussacana** [comment]
same error here.

**Comment by OlegAckbar** [comment]
Same here

**Comment by Ann1kaB** [comment]
I'm using wayland and the NVIDIA GPU seems to just work even on OpenGL, also using the ldn build too! I just launch with `GDK_BACKEND=x11 prime-run ryujinx(-ldn)`

**Comment by EeveeEuphoria** [comment]
> I'm using wayland and the NVIDIA GPU seems to just work even on OpenGL, also using the ldn build too! I just launch with `GDK_BACKEND=x11 prime-run ryujinx(-ldn)`

I'm using x11 and it still doesn't work for me in OpenGL, giving me the exact same problem as the initial post here. I don't know what witchcraft you did to get it to work, but it doesn't work for me.

**Comment by Ann1kaB** [comment]
> > I'm using wayland and the NVIDIA GPU seems to just work even on OpenGL, also using the ldn build too! I just launch with `GDK_BACKEND=x11 prime-run ryujinx(-ldn)`
> 
> I'm using x11 and it still doesn't work for me in OpenGL, giving me the exact same problem as the initial post here. I don't know what witchcraft you did to get it to work, but it doesn't work for me.

the secret sauce is wayland :)

**Comment by EeveeEuphoria** [comment]
> the secret sauce is wayland :)
I actually installed the plasma wayland package just to see if you're right. You're somehow right, what the heck.
Though, I am now unfortunately finding out wayland is really laggy for some reason, so, my dreams of running this in OpenGL will have to wait until I find out why wayland is lagging out.

**Comment by Ann1kaB** [comment]
did some further tinkering and discovered mangohud fps_limit=60 and setting vsync in the emulator OFF fixes the horrible tearing :D

**Comment by marcussacana** [comment]
It's strange how a bug like that still not fixed after so much time, when happens the emulator is literally broken.
In my case I just use Vulkan instead, but is good to see someone find a way to use the opengl
